src/repair.rs

```diff
@@ -93,164 +100,223 @@ impl<N: Network> Repair<N> {
+        let mut repair_receiver = self.repair_channel.1.take().unwrap();
+        loop {
+            let next_timeout = self.request_timeouts.peek().map(|t| t.0);
+            let sleep_duration = match next_timeout {
+                None => std::time::Duration::MAX,
+                Some(t) => t.duration_since(Instant::now()),
+            };
+            tokio::select! {
+                // handle repair request or response from network
+                res = self.receive() => self.handle_repair_message(res.unwrap()).await.unwrap(),
+                // handle request for repairing new block
+                Some(block_id) = repair_receiver.recv() => {
+                    self.repair_block(block_id).await;
+                }
+                // handle next request timeout
+                _ = tokio::time::sleep(sleep_duration) => {
+                    let Some((_, hash)) = self.request_timeouts.pop() else {
+                        continue;
+                    };
+                    if let Some(request) = self.outstanding_requests.remove(&hash) {
+                        debug!("retrying timed-out repair request {request:?}");
+                        self.send_request(request).await.unwrap();
+                    }
+                }
+            }
         }
     }
 
     /// Starts repair process for the block specified by `slot` and `block_hash`.
-    pub async fn repair_block(&self, slot: Slot, block_hash: Hash) {
-        if self
-            .blockstore
-            .read()
-            .await
-            .get_block(slot, block_hash)
-            .is_some()
-        {
+    pub async fn repair_block(&mut self, block_id: BlockId) {
+        let (slot, block_hash) = block_id;
+        let h = &hex::encode(block_hash)[..8];
+        if self.blockstore.read().await.get_block(block_id).is_some() {
+            trace!("ignoring repair for block {h} in slot {slot}, already have the block");
             return;
         }
 
-        let h = &hex::encode(block_hash)[..8];
         debug!("repairing block {h} in slot {slot}");
-        // TODO: perform actual repair
-        // HACK: magic number of 10 requests (to ensure it can handle some failures)
-        for _ in 0..10 {
-            self.request_parent(slot, block_hash).await.unwrap();
+        let req = RepairRequest::LastSliceRoot(block_id);
+        self.send_request(req).await.unwrap();
+    }
+
+    async fn handle_repair_message(&mut self, msg: RepairMessage) -> Result<(), NetworkError> {
+        match msg {
+            RepairMessage::Request(request, sender) => {
+                self.answer_request(request, sender).await?;
+            }
+            RepairMessage::Response(resposne) => {
+                self.handle_response(resposne).await;
+            }
         }
+        Ok(())
     }
 
     /// Tries to answer the given repair request.
     ///
     /// If we do not have the block the request refers to, the request is ignored.
```

Comment:
Again, probably future work.  If we ignore requests that we cannot satisfy, the the requester will have to wait till timeout to try again.  Does it make sense to send a response so that they can retry faster?